New package: CTBase v0.11.2

JuliaRegistrator opened the pull request:
- Registering package: CTBase
- Repository: https://github.com/control-toolbox/CTBase.jl
- Created by: @ocots
- Version: v0.11.2
- Commit: 149094472b5b3d00385d3546320b5fa19f05fe6e
- Reviewed by: @ocots
- Reference: https://github.com/control-toolbox/CTBase.jl/issues/194#issuecomment-2210521017
- Description: Fundamentals of the control-toolbox ecosystem
<!-- bf0c69308befbd3ccf2cc956ac8a4671113b158fd87a6d4e741b0173dab051a48f7dc43e593ffff4357bc647a3a83598618f65af88c1b366ac6be6ae32abc86de15de647b752930376b27dc86f09c71a97644f2d3b624f8826f01241c48c9cada01814a0db40649d95a2ee0c876d6906a86432b7d308cc6e8165555f2bee6c54ca3f1f88b80758c2a8ef2a202dd15bf26c3139692ef7791dc363899f37d158d8c2be1049bda96f6d39933dbb28c40342 -->

ocots commented:
Dear Admin,

Mainly with @jbcaillau, @PierreMartinon and @joseph-gergaud, we develop an optimal control toolbox :

- [GitHub repo](https://github.com/control-toolbox)
- [Documentation](https://control-toolbox.org/docs/optimalcontrol/stable/)

This toolbox is composed of several packages:

```mermaid
flowchart TD
O(<a href='https://control-toolbox.org/docs/optimalcontrol/stable/'>OptimalControl</a>) --> B(<a href='https://control-toolbox.org/docs/ctbase/stable/'>CTBase</a>)
O --> D(<a href='https://control-toolbox.org/docs/ctdirect/stable/'>CTDirect</a>)
O --> F(<a href='https://control-toolbox.org/docs/ctflows/stable/'>CTFlows</a>)
P(<a href='https://control-toolbox.org/docs/ctproblems/stable/'>CTProblems</a>) --> F
P --> B
F --> B
D --> B
```

We intend if accepted to switch from our [local registry](https://github.com/control-toolbox/ct-registry) to the general. We should move `CTBase.jl`, `CTDirect.jl`, `CTFlows.jl` and the main package `OptimalControl.jl`.

Note tha we have presented it at the last JuliaCon: see the [presentation](https://control-toolbox.org/docs/optimalcontrol/v0.8/juliacon2023.html).

goerz commented:
[noblock] Looks fine, but you have to edit your comment to contain `[noblock]`

ocots commented:
> [noblock] Looks fine, but you have to edit your comment to contain `[noblock]`

Thanks. I do it. This means that `CTBase`is ok and we do not have to change the name?

ocots commented:
Here is my new comment: https://github.com/control-toolbox/CTBase.jl/issues/194#issuecomment-2210521017

It doesn't seem to work. @goerz Could you help me?

ocots commented:
[noblock] Ok thanks.

JuliaRegistrator committed:
New package: CTBase v0.11.2

UUID: 54762871-cc72-4466-b8e8-f6c8b58076cd
Repo: https://github.com/control-toolbox/CTBase.jl.git
Tree: 0c060589120ed61765e3a0c1c6a9fdbeda22dfc9

Registrator tree SHA: 17aec322677d9b81cdd6b9b9236b09a3f1374c6a

jbcaillau commented:
[noblock] hi @goerz One question regarding the package names in the hierarchy depicted [above](https://github.com/JuliaRegistries/General/pull/110480#issuecomment-2210531425): I see that you have added the badge below

![IMG_3310](https://github.com/JuliaRegistries/General/assets/62183989/071f7550-50fc-43dc-97cc-5db239693078)


Does this mean that *all* the 4 names in there (`CTxxxx.jl`) are OK? (Would be nice 🤞🏾)

goerz commented:
[noblock] As far as I'm concerned, yes. They still might need overrides for their own registrations, but I'll be happy to apply those. These are not user-facing packages, so I wouldn't put too much scrutiny on naming issues. And having a common prefix is good, even if that prefix wouldn't necessarily fly if those packages were user-facing.

GunnarFarneback commented:
[noblock]
Yeah, in my world CT most definitely means Computed Tomography.

jbcaillau commented:
[noblock] one question @goerz : we were planning to have `OptimalControl.jl` registered for JuliaCon next week (guess we're not alone with such demands, though 😓) - talk planned Wednesday 10. We were actually not aware of the three day mandatory waiting period. According to the graph of dependencies below, involving packages from a [local register](https://github.com/control-toolbox/ct-registry), I suppose that the normal process is to
- wait for `CTBase.jl` to be registered
- ask for both `CTDirect.jl` and `CTFlows.jl` to be registered
- eventually ask for `OptimalContorl.jl` to be registered.

Which, according to the same rule, should take  3 + 3 = 6 more days: any chance to speed the process up?

NB. I suppose that starting a registration process for `OptimalControl.jl` that currently still has dependencies towards packages in a local registry is useless, since registration checks would not pass.

> ```mermaid
> flowchart TD
> O(<a href='https://control-toolbox.org/docs/optimalcontrol/stable/'>OptimalControl</a>) --> B(<a href='https://control-toolbox.org/docs/ctbase/stable/'>CTBase</a>)
> O --> D(<a href='https://control-toolbox.org/docs/ctdirect/stable/'>CTDirect</a>)
> O --> F(<a href='https://control-toolbox.org/docs/ctflows/stable/'>CTFlows</a>)
> F --> B
> D --> B
> ```

goerz commented:
> I suppose that starting a registration process for OptimalControl.jl that currently still has dependencies towards packages in a local registry is useless, since registration checks would not pass.

I think you can start the registration process now, and then just retrigger when the dependencies are merged. That’s probably the best way to speed things up [noblock]

goerz commented:
[noblock] It might also be possible to prepare a manual PR that registers all the org-packages in one go. That’ll definitely require a manual merge from a full maintainer, though (I don’t have merge permissions). I’m not sure what the maintainers’ availability is going to be, especially during JuliaCon, so this might not be faster.

goerz commented:
[noblock] I really think we should have a rule that packages have to be registered at the time a JuliaCon abstract is accepted. Personally, I’m quite opposed to circumventing the waiting period. It’s rather unfair to other submissions. Maybe even especially for occasions like JuliaCon, since a lot of people will be traveling and new submissions are likely to get less systematically reviewed than at other times.

GunnarFarneback commented:
[noblock]
> I think you can start the registration process now, and then just retrigger when the dependencies are merged. That’s probably the best way to speed things up

Yes, this is the way to go, since it gives the normal vetting time for the names and let the automerge checks run properly.

To be clear, when you register a package with a not yet registered dependency, automerge will fail to load the package, and say so. But once the dependency is registered, you can re-register the dependee. Then the automerge checks will be restarted and should now be able to load the package. As long as you don't change the version (or URL), the original PR will be updated and the timer will not be reset.

jbcaillau commented:
[noblock] hi @goerz @GunnarFarneback thanks for the feedback. *dura lex, sed lex*: fine with me. and JuliaCon period is definitely not the right time to ask for a manual PR. 

So we follow the advice and initiate the process for `OptimalControl.jl` and the rest of the `CTxxxx` local registry. @ocots you're in gear 🙂.

> 
> > I think you can start the registration process now, and then just retrigger when the dependencies are merged. That’s probably the best way to speed things up
> 
> Yes, this is the way to go, since it gives the normal vetting time for the names and let the automerge checks run properly.
> 
> To be clear, when you register a package with a not yet registered dependency, automerge will fail to load the package, and say so. But once the dependency is registered, you can re-register the dependee. Then the automerge checks will be restarted and should now be able to load the package. As long as you don't change the version (or URL), the original PR will be updated and the timer will not be reset.

ocots commented:
[noblock] I have started the process for `CTFlows.jl`, `CTDirect.jl` and `OptimalControl.jl`:

- https://github.com/control-toolbox/CTFlows.jl/issues/27
- https://github.com/control-toolbox/CTDirect.jl/issues/156
- https://github.com/control-toolbox/OptimalControl.jl/issues/208

```mermaid
flowchart TD
O(<a href='https://control-toolbox.org/docs/optimalcontrol/stable/'>OptimalControl</a>) --> B(<a href='https://control-toolbox.org/docs/ctbase/stable/'>CTBase</a>)
O --> D(<a href='https://control-toolbox.org/docs/ctdirect/stable/'>CTDirect</a>)
O --> F(<a href='https://control-toolbox.org/docs/ctflows/stable/'>CTFlows</a>)
F --> B
D --> B
```

jbcaillau commented:
[noblock] @goerz specifically on this: disagree 🙂. having a local registry (instead of registering in general) makes sense in many cases, JuliaCon or not.  

> I really think we should have a rule that packages have to be registered at the time a JuliaCon abstract is accepted.

jbcaillau commented:
[noblock] A very timely message from JuliaCon 🙂
> 
> [Juliacon 2024](https://pretalx.com/juliacon2024/)
> Package registration around JuliaCon
>
> Dear Jean-Baptiste Caillau, Alesia Herasimenka, Olivier Cots,
>
> If you are planning to register a package related to your upcoming JuliaCon 2024 presentation Trajectory optimisation in space mechanics with Julia please be aware of a few things related to package registration:
> 
> Package registration has an automatic 72 hour wait period. Except in extreme extenuating circumstances this wait period is mandatory and may not be shortened or skipped. A JuliaCon talk is not typically a sufficient reason to override.
>
> Package names are subject to certain guidelines and limitations, along with community input. If you do plan to register a package related to your talk we strongly encourage you to make that registration now to give time for community feedback before your presentation.
>
> A talk or paper is not sufficient reason for a package naming override. If your name is too similar to an existing project or fails to meet some other guidelines it may be rejected even if you have already published content related to your package using that name. Existing published content may be a good but insufficient reason to override naming guidelines, this is handled on a case by case basis and cannot be relied upon.
>
> If your package is not registered in time for your talk you can always encourage people to install it using the URL rather than simply the name.
>
> If you have multiple highly related packages (for instance a base/interface package and a core package) you may submit all registration pull requests at the same time. Failed dependencies will prevent some of those packages from being green on CI, however it will allow community feedback so you incur the 3 day wait period once rather than 3 days * #number of packages you are registering.
>
> We're excited to see you in Eindhoven next week!
> Sincerely,
> Raye Kimmerer
> JuliaCon 2024 Executive Chair

goerz commented:
[noblock] Well, that message was prompted by this particular PR ;-) (`#community` channel on the Julia Slack)

ocots commented:
[noblock] @goerz any news for this merge 🙂?

goerz commented:
[noblock] Very strange. It should have been auto-merged. Is there a comment that is missing `[noblock]` (or misspells it)? I'm not seeing the problem.

goerz commented:
[noblock] I asked on Slack for someone to manually click the merge button, in case we can't figure out why it didn't merge. Not sure if there are a lot of maintainers available in the middle of JuliaCon, though. 🤞

jbcaillau commented:
[noblock] thanks @goerz never mind, though. right now we're enjoying the Aerospace MS, so everything's fine 🙂

giordano commented:
I think there's a github issue here, I can't merge nor rebase the branch.

[noblock]

giordano commented:
Ok, I changed the target branch from `master` to `master` (yes, you read it right), let's see if that helps anything.

[noblock]

giordano commented:
Apparently that was it

giordano commented:
> If it helps, we can close this PR and redo the process.

I'm moderately sure that just re-registering the version without closing the PR would have worked, too, as it would have updated the base branch.

jbcaillau commented:
🙏🏽 @goerz @giordano thank you guys: going `general` exactly 4 minutes before our [talk](https://control-toolbox.org/docs/optimalcontrol/stable/juliacon2024.html)  this is scheduling 🙂

now we just have to wait a few more days for the rest of the `CT`-hierarchy 🤞🏾

ocots commented:
@goerz @giordano Many thanks!

I have started the process for the rest.

`CTDirect.jl` is 🆗 : https://github.com/JuliaRegistries/General/pull/110817
`CTflows.jl` is 🆗 : https://github.com/JuliaRegistries/General/pull/110818

When it is finished, there will be to register the main package `OptimalControl.jl`!
